#!/bin/bash
set -e -o pipefail

#######################################
# load environment variables
source ./../../env.sh
PATH=$ucsc_dir:$PATH
#######################################
# set project-specific variables
batch_id="Batch_Example" # The batch_id used for the processing batch. Default = "Batch_Example".
threads=4 # The number of threads to use. Default = 4.
contrast="vir1,VIRc" # format: test_group_tag,control_group_tag. Must be the same as the name in the comparison_group column of the master_sample_table.
read_counts_cutoff=5 # The read count cutoff for the differentially expressed genes/isoforms table. Default = 5.
log2foldchange_cutoff=1 # The log2 foldchange cutoff for the differential expression volcano plot. Default = 1.
adj_p_value_cutoff=0.05 # The adjusted p-value cutoff for the differential expression volcano plot. Default = 0.05.
interested_genes="FLC"  # Interested genes (comma separated gene names, eg., FLC,CCA1) would like to be highlighted in volcano plots. Default = NULL.
debug="no" # Whether to keep intermediate files for debuging. Use "yes" if prefer to keep intermediate files, otherwise use "no". Default = "no".
############################################################
# Normally no need to change the following settings.
    quant_table="./../02.Isoform_Clustering_and_Quantification/${batch_id}/all_samples_combined/${batch_id}.all_samples_combined.counts_matrix.tidy.txt"  # The isoform count table generated by Module 02. 
raw_quant_table="./../02.Isoform_Clustering_and_Quantification/${batch_id}/all_samples_combined/${batch_id}.all_samples_combined.counts_matrix.tsv"
    isoform_bed="./../02.Isoform_Clustering_and_Quantification/${batch_id}/all_samples_combined/${batch_id}.all_samples_combined.flair_all_collapsed.isoforms.bed"
diffsplice_outdir="./${batch_id}/all_samples_combined/${batch_id}_differential_splicing_output"
##########################################################
echo "Detecting genes/isoforms with differential expression ..."
Rscript --vanilla $NANOTRANS_HOME/scripts/differential_expression.R \
    --tidy_count_table ${quant_table} \
    --read_counts_cutoff ${read_counts_cutoff} \
    --log2foldchange_cutoff ${log2foldchange_cutoff} \
    --adj_p_value_cutoff ${adj_p_value_cutoff} \
    --contrast ${contrast} \
    --interested_genes ${interested_genes} \
    --batch_id ${batch_id} 

source $miniconda3_dir/activate ../../build/flair_conda_env/

echo "Detecting isoforms with differential usage ..."
Rscript --vanilla $NANOTRANS_HOME/scripts/differential_isoform_usage.R \
    --tidy_count_table ${quant_table} \
    --contrast ${contrast} \
    --batch_id ${batch_id} 

echo "Detecting differential splicing events ..."

[[ -d ${diffsplice_outdir} ]] || mkdir -p ${diffsplice_outdir}

call_ds_program=${build_dir}/flair_conda_env/lib/python*/site-packages/flair
python ${call_ds_program}/call_diffsplice_events.py ${isoform_bed} ${diffsplice_outdir}/diffsplice ${raw_quant_table}
python ${call_ds_program}/es_as.py ${isoform_bed} > ${diffsplice_outdir}/diffsplice.es.events.tsv
python ${call_ds_program}/es_as_inc_excl_to_counts.py ${raw_quant_table} ${diffsplice_outdir}/diffsplice.es.events.tsv > ${diffsplice_outdir}/diffsplice.es.events.quant.tsv

Rscript --vanilla $NANOTRANS_HOME/scripts/differential_splicing.R \
    --inputs_dir ${diffsplice_outdir} \
    --contrast ${contrast} \
    --threads ${threads} \
    --batch_id ${batch_id} 

source $miniconda3_dir/deactivate

if [[ ${debug} == "no" ]]; then
  rm ${diffsplice_outdir}/diffsplice.*.events.tsv
fi

############################
# checking bash exit status
if [[ $? -eq 0 ]]
then
    echo ""
    echo "#########################################################################"
    echo ""
    echo "NanoTrans message: This bash script has been successfully processed! :)"
    echo ""
    echo "#########################################################################"
    echo ""
    exit 0
fi
############################
